src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/SignupController.java

```diff
@@ -0,0 +1,66 @@
+package dk.cngroup.trainings.spring.springassignment.controller.web;
+
+import dk.cngroup.trainings.spring.springassignment.controller.dto.user.UserEntityDtoConverter;
+import dk.cngroup.trainings.spring.springassignment.controller.dto.user.UserSignupDTO;
+import dk.cngroup.trainings.spring.springassignment.exception.UserEmailExistsException;
+import dk.cngroup.trainings.spring.springassignment.exception.UserPasswordsNotMatchingException;
+import dk.cngroup.trainings.spring.springassignment.model.User;
+import dk.cngroup.trainings.spring.springassignment.service.helper.UserDtoValidatorService;
+import dk.cngroup.trainings.spring.springassignment.service.user.UserService;
+import org.springframework.stereotype.Controller;
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.ModelAttribute;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.servlet.ModelAndView;
+
+import javax.validation.Valid;
+
+@Controller
+@RequestMapping("/signup")
+public class SignupController {
+
+	private final String VIEW_PATH = "views/authentication/";
+
+	private UserService userService;
+
+	public SignupController(UserService userService) {
+		this.userService = userService;
+	}
+
+	@GetMapping("")
+	public ModelAndView signupForm() {
+		ModelAndView modelAndView = new ModelAndView(VIEW_PATH + "signup");
+		modelAndView.addObject("userSignupDto", new UserSignupDTO());
+		return modelAndView;
+	}
+
+	@PostMapping("")
+	public ModelAndView signup(@ModelAttribute("userSignupDto") @Valid UserSignupDTO userSignupDTO,
+							   BindingResult bindingResult) {
+		if (bindingResult.hasErrors()) {
+			ModelAndView modelAndView = new ModelAndView(VIEW_PATH + "signup");
+			return modelAndView;
+		} else {
+			try {
+				UserDtoValidatorService.checkMatchingPasswords(userSignupDTO);
```

Comment:
regardin you question 1 in Skype: yes, this should probably be moved to the service layer, that is too much logic for controllers.

So maybe you can create a new service just for this (and maybe something else), which you will call. And this service will throw only one type of exception, which will include all information you need in Controller. So inside the service, you can catch these exceptions and wrap them in exception of suitable type, which you can catch in the controller and use it for error messages.

Reply:
or it does not have to be service exactly (because you are working with DTO, which is controller level)... it can be some other class in controller package to perform required operations

Reply:
So I only need to move that class to the controller package?

Reply:
Now, but you can create new class, which will handle this logic. This class will probably be in controller package. But in this class you shoud do the Exception wrapping, so you don't have to deal with so many try and catch here in the controller.